neurom/core/_soma.py

```
@@ -202,84 +208,34 @@ class SomaSimpleContour(Soma):
 
-    if(npoints == 3 and
-       points[0][COLS.P] == -1 and
-       points[1][COLS.P] == 1 and
-       points[2][COLS.P] == 1):
-        warnings.warn('Using neuromorpho 3-Point soma')
-        # NeuroMorpho is the main provider of morphologies, but they
-        # with SWC as their default file format: they convert all
-        # uploads to SWC.  In the process of conversion, they turn all
-        # somas into their custom 'Three-point soma representation':
-        #  http://neuromorpho.org/SomaFormat.html
 
-        return SomaNeuromorphoThreePointCylinders
-
-    return {0: None,
-            1: SomaSinglePoint}.get(npoints, SomaCylinders)
-
-
-def make_soma(points, soma_check=None, soma_class=SOMA_CONTOUR):
+def make_soma(morphio_soma):
     """Make a soma object from a set of points.
 
-    Infers the soma type (SomaSinglePoint, SomaSimpleContour)
-    from the points and the 'soma_class'
-
-    Arguments:
-        points: collection of points forming a soma.
-        soma_check: optional validation function applied to points. Should
-        raise a SomaError if points not valid.
-        soma_class(str): one of 'contour' or 'cylinder' to specify the type
-
-    Raises:
-        SomaError if no soma points found, points incompatible with soma, or
-        if soma_check(points) fails.
+    Args:
+        morphio_soma(morphio.Soma): soma instance of MorphIO
     """
-    if soma_check:
-        soma_check(points)
-
-    stype = _get_type(points, soma_class)
-
-    if stype is None:
-        raise SomaError('Invalid soma points')
-
-    return stype(points)
+    SomaBuilders = {
+        SomaType.SOMA_SINGLE_POINT: SomaSinglePoint,
+        SomaType.SOMA_CYLINDERS: SomaCylinders,
+        SomaType.SOMA_NEUROMORPHO_THREE_POINT_CYLINDERS: SomaNeuromorphoThreePointCylinders,
+        SomaType.SOMA_SIMPLE_CONTOUR: SomaSimpleContour,
+    }
+
+    return SomaBuilders.get(morphio_soma.type, SomaSimpleContour)(morphio_soma)
```

I guess you should raise if the soma.type is unknown

---

Do you mean to raise if soma type is `SomaType.SOMA_UNDEFINED` because I see that such case isn't processed by NeuroM

---

Imagine we add a `SomaType.NEW_SOMA_TYPE` in morphio. If the dispatch dict is not up to date, you will silently return a SomaSimpleContour which is a bit shaky.

---

done